architecture/2025-10-terraform-bicep-settings.md

```diff
@@ -0,0 +1,325 @@
+- Terraform executor resolves versioned binary paths, renders `.terraformrc`, configures backends, and emits structured logs.
+
+### Architecture Diagram
+
+```mermaid
+flowchart LR
+    subgraph Operator Workstation
+        CLI[rad CLI]
+    end
+    subgraph Kubernetes Cluster
+        subgraph Radius Control Plane
+            API[Installer REST Endpoint]
+            QUEUE["Installer Queue\n(QueueMessage CR)"]
+            CTRL["Installer Async Handler"]
+            ENVCTRL[Environment Controller]
+            EXEC[Terraform Executor]
+        end
+        PVC[(Terraform Binary PVC)]
+        SETTINGS[terraformSettings / bicepSettings]
+    end
+
+    CLI -->|install/uninstall/status| API
+    API -->|enqueue job| QUEUE
+    QUEUE -->|lease job| CTRL
+    CTRL -->|download & verify| PVC
+    CTRL -->|persist metadata| API
+    ENVCTRL --> SETTINGS
+    ENVCTRL --> EXEC
+    EXEC -->|reads binaries| PVC
+    SETTINGS --> ENVCTRL
+```
+
+### Detailed Design
+
+- TypeSpec adds `Radius.Core/terraformSettings@2025-08-01-preview` and `Radius.Core/bicepSettings@2025-08-01-preview` aligned with the feature spec.
+- The `Radius.Core/environments` controller sources Terraform/Bicep inputs exclusively from the new settings resources, while the legacy `Applications.Core/environments` controller continues serving `recipeConfig` callers until deprecation. Both paths reuse the existing air-gapped download/auth/TLS helpers so we avoid duplicating plumbing.
+- Installer async handler (plugged into the existing worker pipeline in `pkg/server/asyncworker.go`) downloads Terraform binaries, verifies checksums, persists version metadata (requested URL, checksum, install timestamp, health) in installer status storage, and places binaries on the shared PVC mount (for example `/mnt/radius-terraform/<version>/`). Its queue is configured with `MaxOperationConcurrency = 1`, so install/uninstall jobs execute strictly in submission order; uninstall only removes a version once no executions reference it.
+- CLI invokes installer APIs for install/uninstall/status; Phase 2 may add a `validate` call that reuses the same infrastructure when preflight checks are implemented.
+- TerraformSettings serializer covers `.terraformrc` (provider mirrors, credentials, env vars) and backend blocks. Phase 1 keeps the existing Kubernetes backend adapter; AzureRM/S3 adapters plug in during Phase 2, with other backends passed through without managed auth until prioritized.
+- Migration guidance stays in the Phase 2 backlog: controllers emit warnings while `recipeConfig` remains, and documentation will walk operators through removing the legacy fields.
+- Recipe execution resolves the pinned binary path via the stored metadata, preserving multi-tenant isolation so environments can run different Terraform versions without interference.
```

Comment:
Do we need to pause recipe deployments while Terraform installation operations are running? This would obviously not be a problem on small clusters with few users, but could be a problem at scale. 

On the other hand, maybe we should defer synchronization to look at all the scenarios in which we may need to serialize operations. Are there other scenarios?